simapp/sim_test.go

```diff
@@ -45,120 +44,186 @@ func init() {
+				stakingtypes.HistoricalInfoKey, stakingtypes.UnbondingIDKey, stakingtypes.UnbondingIndexKey,
+				stakingtypes.UnbondingTypeKey,
+			},
+			authzkeeper.StoreKey:   {authzkeeper.GrantQueuePrefix},
+			feegrant.StoreKey:      {feegrant.FeeAllowanceQueueKeyPrefix},
+			slashingtypes.StoreKey: {slashingtypes.ValidatorMissedBlockBitmapKeyPrefix},
+		}
+		AssertEqualStores(t, app, newApp, app.SimulationManager().StoreDecoders, skipPrefixes)
+	})
+}
 
-	// Run randomized simulation
-	_, simParams, simErr := simulation.SimulateFromSeed(
-		t,
-		os.Stdout,
-		app.BaseApp,
-		simtestutil.AppStateFn(app.AppCodec(), app.AuthKeeper.AddressCodec(), app.StakingKeeper.ValidatorAddressCodec(), app.SimulationManager(), app.DefaultGenesis()),
-		simtypes.RandomAccounts, // Replace with own random account function if using keys other than secp256k1
-		simtestutil.SimulationOperations(app, app.AppCodec(), config),
-		BlockedAddresses(),
-		config,
-		app.AppCodec(),
-		codectestutil.CodecOptions{}.GetAddressCodec(),
-	)
+// Scenario:
+//
+//	Start a fresh node and run n blocks, export state
+//	set up a new node instance, Init chain from exported genesis
+//	run new instance for n blocks
+func TestAppSimulationAfterImport(t *testing.T) {
+	sims.Run(t, NewSimApp, setupStateFactory, func(t *testing.T, ti sims.TestInstance[*SimApp]) {
+		app := ti.App
+		t.Log("exporting genesis...\n")
+		exported, err := app.ExportAppStateAndValidators(false, []string{}, []string{})
+		require.NoError(t, err)
+
+		t.Log("importing genesis...\n")
+		newTestInstance := sims.NewSimulationAppInstance(t, ti.Cfg, NewSimApp)
+		newApp := newTestInstance.App
+		_, err = newApp.InitChain(&abci.InitChainRequest{
+			AppStateBytes: exported.AppState,
+			ChainId:       sims.SimAppChainID,
+		})
+		if IsEmptyValidatorSetErr(err) {
+			t.Skip("Skipping simulation as all validators have been unbonded")
+			return
+		}
+		require.NoError(t, err)
+		newStateFactory := setupStateFactory(newApp)
+		_, err = simulation.SimulateFromSeedX(
+			t,
+			newTestInstance.AppLogger,
+			sims.WriteToDebugLog(newTestInstance.AppLogger),
+			newApp.BaseApp,
+			newStateFactory.AppStateFn,
+			simtypes.RandomAccounts,
+			simtestutil.SimulationOperations(newApp, newApp.AppCodec(), newTestInstance.Cfg, newApp.TxConfig()),
+			newStateFactory.BlockedAddr,
+			newTestInstance.Cfg,
+			newStateFactory.Codec,
+			newApp.TxConfig().SigningContext().AddressCodec(),
+			ti.ExecLogWriter,
+		)
+		require.NoError(t, err)
+	})
+}
 
-	// export state and simParams before the simulation error is checked
-	err := simtestutil.CheckExportSimulation(app, config, simParams)
-	require.NoError(t, err)
-	require.NoError(t, simErr)
+func IsEmptyValidatorSetErr(err error) bool {
+	return err != nil && strings.Contains(err.Error(), "validator set is empty after InitGenesis")
+}
 
-	if config.Commit {
-		simtestutil.PrintStats(db)
+func TestAppStateDeterminism(t *testing.T) {
+	const numTimesToRunPerSeed = 3
```

Comment:
Maybe this value can be externally set, for example CI for releases could use a higher value for a more "secure" coverage.

Reply:
The 3 was in the original code. Not sure if a higher value is the best way for a higher probability of collusions. You can also argue, adding more TX or blocks increases the probability. 
I don't have a strong opinion but I want to avoid adding another CLI flag in this PR. Let's iterate on this.